### PR TITLE
Update PromQL upto Jan-20 (@ <timestamp> addition)

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen-Singh <harkishensingh@hotmail.com>

TODO:

- [x] update engine_test.go
- [x] add @ based tests

### CODE REVIEW

**Comment by ldez** [comment]
https://github.com/golangci/golangci-lint/issues/827#issuecomment-765991719

The root cause of your problem is the upgrade of `github.com/docker/docker` is your PR.
```diff
- github.com/docker/docker v17.12.0-ce-rc1.0.20200706150819-a40b877fbb9e+incompatible h1:+mzU0jHyjWpYHiD0StRlsVXkCvecWS2hc55M3OlUJSk=
- github.com/docker/docker v17.12.0-ce-rc1.0.20200706150819-a40b877fbb9e+incompatible/go.mod h1:eEKB0N0r5NX/I1kEveEz05bcu8tLC/8azJZsviup8Sk=
+ github.com/docker/docker v20.10.1+incompatible h1:u0HIBLwOJdemyBdTCkoBX34u3lb5KyBo0rQE3a5Yg+E=
+ github.com/docker/docker v20.10.1+incompatible/go.mod h1:eEKB0N0r5NX/I1kEveEz05bcu8tLC/8azJZsviup8Sk=
```

As you are using `github.com/testcontainers/testcontainers-go@v0.5.0` you have to propagate manually the `require` directives from this project:
https://github.com/testcontainers/testcontainers-go/blob/v0.5.0/go.mod#L37

The `require` directives are not transitive, you have to handle them by hand.

In summary, your `go.mod` must be updated like that:

```
module github.com/timescale/promscale

//TODO remove after repo rename
replace github.com/timescale/promscale => ./

go 1.14

require (
	github.com/NYTimes/gziphandler v1.1.1
	github.com/blang/semver/v4 v4.0.0
	github.com/cespare/xxhash/v2 v2.1.1
	github.com/containerd/containerd v1.4.1 //indirect
	github.com/docker/go-connections v0.4.0
	github.com/edsrzf/mmap-go v1.0.0
	github.com/go-kit/kit v0.10.0
	github.com/gogo/protobuf v1.3.1
	github.com/golang/snappy v0.0.2
	github.com/inhies/go-bytesize v0.0.0-20201103132853-d0aed0d254f8
	github.com/jackc/pgconn v1.6.3
	github.com/jackc/pgerrcode v0.0.0-20190803225404-afa3381909a6
	github.com/jackc/pgproto3/v2 v2.0.2
	github.com/jackc/pgtype v1.4.2
	github.com/jackc/pgx/v4 v4.8.0
	github.com/moby/sys/mount v0.2.0 // indirect
	github.com/opentracing-contrib/go-stdlib v1.0.0
	github.com/opentracing/opentracing-go v1.2.0
	github.com/peterbourgon/ff/v3 v3.0.0
	github.com/pkg/errors v0.9.1
	github.com/prometheus/client_golang v1.9.0
	github.com/prometheus/client_model v0.2.0
	github.com/prometheus/common v0.15.0
	github.com/prometheus/prometheus v1.8.2-0.20210121114440-a7e446cf2d2a
	github.com/schollz/progressbar/v3 v3.7.2
	github.com/stretchr/testify v1.6.1
	github.com/testcontainers/testcontainers-go v0.5.1
	github.com/uber/jaeger-client-go v2.25.0+incompatible
	go.uber.org/atomic v1.7.0
	go.uber.org/goleak v1.1.10
	golang.org/x/sys v0.0.0-20210122093101-04d7465088b8 // indirect
)

replace github.com/docker/docker => github.com/docker/engine v0.0.0-20190717161051-705d9623b7c1
```

The problem is `github.com/docker/docker` because:
- they renamed their organization
- they "splitter" their repositories
- to get `github.com/docker/docker` you have to use `github.com/docker/engine`
- they don't use go modules.

Welcome to the Docker dependencies hell :smile:

**Comment by Harkishen-Singh** [comment]
Hey @ldez  thank you for your reply here. The explanation is awesome and I will go through it soon and see if the problem gets resolved. 🙏

File: go.sum [hunk 1]
```diff
@@ -16,6 +16,7 @@ cloud.google.com/go v0.56.0/go.mod h1:jr7tqZxxKOVYizybht9+26Z/gUq7tiRzu+ACVAMbKV
 cloud.google.com/go v0.57.0/go.mod h1:oXiQ6Rzq3RAkkY7N6t3TcE6jE+CIBBbA36lwQ1JyzZs=
```
Comment (cevian) [inline go.sum:16]: can we clean this up by regenerating?

File: pkg/promql/engine.go [hunk 2]
```diff
@@ -651,34 +692,72 @@ func (f inspector) Visit(node parser.Node, path []parser.Node) (parser.Visitor,
 	return f, nil
 }
 
-// This is modified from upstream to avoid descending into a MatrixSelector's VectorSelector
-func (ng *Engine) findMinTime(s *parser.EvalStmt) (time.Time, error) {
-	var maxOffset time.Duration
-
-	err := parser.Walk(inspector(func(node parser.Node, path []parser.Node) (bool, error) {
-		subqOffset, subqRange := ng.subqueryOffsetRange(path)
+func (ng *Engine) findMinMaxTime(s *parser.EvalStmt) (int64, int64) {
```
Comment (cevian) [inline pkg/promql/engine.go:695]: with this change, I believe we can delete the inspector type above (line 685-693)

**Review by cevian — CHANGES_REQUESTED** [review]
Need to block the @ stuff behind a flag just like prometheus does (additional commit).

Also, lets rename the second commit to `Update findMinMaxTime() from upstream (undo 3eae912)`

File: go.sum (hunk 1 above)
Comment (Harkishen-Singh) [reply go.sum:16]: I think the current state is after the `go mod tidy`. Regenerating the mod again might create issues. WDYT?

**Pull request closed without merging**